components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt

```diff
@@ -18,91 +22,255 @@ import android.os.ParcelFileDescriptor
 import mozilla.components.feature.downloads.ext.withResponse
-import mozilla.components.support.base.ids.NotificationIds
+import mozilla.components.support.base.ids.cancel
 import mozilla.components.support.base.ids.notify
 import java.io.File
 import java.io.FileOutputStream
 import java.io.IOException
+import java.io.InputStream
 import java.io.OutputStream
+import kotlin.random.Random
 
 /**
  * Service that performs downloads through a fetch [Client] rather than through the native
  * Android download manager.
  *
  * To use this service, you must create a subclass in your application and it to the manifest.
  */
-abstract class AbstractFetchDownloadService : CoroutineService() {
+@Suppress("TooManyFunctions", "LargeClass")
+abstract class AbstractFetchDownloadService : Service() {
 
     protected abstract val httpClient: Client
     @VisibleForTesting
     internal val broadcastManager by lazy { LocalBroadcastManager.getInstance(this) }
     @VisibleForTesting
     internal val context: Context get() = this
 
-    override fun onCreate() {
-        startForeground(
-            NotificationIds.getIdForTag(context, ONGOING_DOWNLOAD_NOTIFICATION_TAG),
-            DownloadNotification.createOngoingDownloadNotification(context)
-        )
-        super.onCreate()
+    private var downloadJobs = mutableMapOf<Long, DownloadJobState>()
+
+    private data class DownloadJobState(
+        var job: Job? = null,
+        var state: DownloadState,
+        var currentBytesCopied: Long = 0,
+        var status: DownloadJobStatus,
+        var foregroundServiceId: Int = 0
+    )
+
+    private enum class DownloadJobStatus {
+        ACTIVE,
+        PAUSED,
+        CANCELLED,
+        FAILED
+    }
+
+    private val broadcastReceiver by lazy {
+        object : BroadcastReceiver() {
+            override fun onReceive(context: Context, intent: Intent?) {
+                val downloadId =
+                        intent?.extras?.getLong(DownloadNotification.EXTRA_DOWNLOAD_ID) ?: return
+                val currentDownloadJobState = downloadJobs[downloadId] ?: return
+
+                when (intent.action) {
+                    ACTION_PAUSE -> {
+                        currentDownloadJobState.status = DownloadJobStatus.PAUSED
+                        currentDownloadJobState.job?.cancel()
+                    }
+
+                    ACTION_RESUME -> {
+                        NotificationManagerCompat.from(context).cancel(
+                            currentDownloadJobState.foregroundServiceId
+                        )
+
+                        currentDownloadJobState.job = CoroutineScope(IO).launch {
+                            startDownloadJob(currentDownloadJobState.state)
+                        }
+
+                        currentDownloadJobState.status = DownloadJobStatus.ACTIVE
+                    }
+
+                    ACTION_CANCEL -> {
+                        currentDownloadJobState.status = DownloadJobStatus.CANCELLED
```

Comment:
nit: not sure if it's a problem, but it looks suspicious, that we some times set the status before we apply the change and some times (other actions) after?

Reply:
I'm fairly confident the order does not matter. That being said, I made them all consistent so there's no confusion surrounding it.